src/utils/internalLinks.js

```diff
@@ -169,10 +180,14 @@ export const getNarrowFromLink = (
       const ids = parsePmOperand(paths[1]);
       return pmNarrowFromRecipients(pmKeyRecipientsFromIds(ids, ownUserId));
     }
-    case 'topic':
-      return topicNarrow(parseStreamOperand(paths[1], streamsById), parseTopicOperand(paths[3]));
-    case 'stream':
-      return streamNarrow(parseStreamOperand(paths[1], streamsById));
+    case 'topic': {
+      const streamName = parseStreamOperand(paths[1], streamsById, streamsByName);
+      return streamName == null ? null : topicNarrow(streamName, parseTopicOperand(paths[3]));
```

Comment:
> (I think all clients are made aware of all archived streams and private streams, right? If not, that would add more cases.)

No, I believe the client doesn't get to know about private streams the user isn't in, unless they're an admin. See: https://zulip.com/help/stream-permissions#private-streams

But this basically behaves as an example of your first case, because as far as this user can see it's supposed to be as if that stream never existed.



> I think it'd be reasonable to give the user an error message if the link is well-formed but points to an unknown stream. I don't think it's necessarily worth more than a TODO for now, though.

Yeah, that'd be reasonable.



> I guess this change does add another codepath that ends up violating this bit of `getFileTemporaryUrl`'s jsdoc, though:

Hmm, indeed.

That's not any kind of live bug, because that API endpoint has no side effects -- so any request that doesn't produce a useful, intended result will just result in an error. And that's caught by `tryGetFileTemporaryUrl`, and then its caller falls back to using the original URL directly. But it would be nice to clean up, in jsdoc and/or logic.